File: cpp/src/grpc/adaptee.cpp
#include <iostream>
#include "adaptee.h"

GrpcAdaptee::GrpcAdaptee()
{
    cout << "GrpcAdaptee" << endl;
}

GrpcAdaptee::~GrpcAdaptee()
{
    cout << "~GrpcAdaptee" << endl;
}

GrpcSubscriberAdaptee::GrpcSubscriberAdaptee()
{
    cout << "GrpcSubscriberAdaptee" << endl;
}

GrpcSubscriberAdaptee::~GrpcSubscriberAdaptee()
{
    cout << "~GrpcSubscriberAdaptee" << endl;
}

GrpcProviderAdaptee::GrpcProviderAdaptee()
{
    cout << "GrpcProviderAdaptee" << endl;
}

GrpcProviderAdaptee::~GrpcProviderAdaptee()
{
    cout << "~GrpcProviderAdaptee" << endl;
}

ZDServiceSubscriber_SUBSCRIBER_TYPE GrpcSubscriberAdaptee::getSubscriberType() const
{
    return  ZDServiceSubscriber_SUBSCRIBER_TYPE::ZDServiceSubscriber_SUBSCRIBER_TYPE_PROVIDER;
}

ZDServiceSubscriber_SUBSCRIBER_TYPE GrpcProviderAdaptee::getSubscriberType() const
{
    return  ZDServiceSubscriber_SUBSCRIBER_TYPE::ZDServiceSubscriber_SUBSCRIBER_TYPE_OBSERVER;
}

File: node/src/lib/grpc/adaptee.ts
import { ZDServiceRequest, ZDServiceSubscriber, ZDService, ZDServiceRequestResult } from '../proto/manager_pb';
import { SERVICE_TYPEMap } from '../proto/wrappers_pb';

abstract class GrpcAdaptee {
    abstract getSubscriberId(): string;
    abstract getGrpcHost(): string;
    abstract getSubscriberServiceTypes(): SERVICE_TYPEMap[keyof SERVICE_TYPEMap][];
}

export abstract class GrpcSubscriberAdaptee extends GrpcAdaptee {
    abstract onSubscriberServiceRequest(req: ZDServiceRequest): void;
    public subscriberType: ZDServiceSubscriber.SUBSCRIBER_TYPEMap[keyof ZDServiceSubscriber.SUBSCRIBER_TYPEMap];
    constructor() {
        super();
        this.subscriberType = ZDServiceSubscriber.SUBSCRIBER_TYPE.OBSERVER;
    }
}

export abstract class GrpcProviderAdaptee extends GrpcAdaptee {
    abstract onSubscriberServiceRequest(req: ZDServiceRequest): ZDServiceRequestResult | null;
    abstract getCurrentServiceStatus(): ZDService[];
    public subscriberType: ZDServiceSubscriber.SUBSCRIBER_TYPEMap[keyof ZDServiceSubscriber.SUBSCRIBER_TYPEMap];
    constructor() {
        super();
        this.subscriberType = ZDServiceSubscriber.SUBSCRIBER_TYPE.PROVIDER;
    }
}